coursera/downloaders.py

```diff
@@ -46,6 +49,18 @@ def download(self, url, filename, resume=False):
         is aborted by the user, the partially downloaded file is also removed.
         """
 
+        logging.debug('"raw" filename: <%s>.', filename)
+        filename = clean_filename(filename, minimal_change=True, keep_slashes=True)
+        logging.debug('filename after cleaning: <%s>.', filename)
+
+        # FIXME (rbrito): This should, perhaps, be included clean_filename
+        # m = re.match(r'^(.+)\?(.+=.+&?)*$', filename)
+        m = re.match(r'^([^?#]+)[?#]?.*$', filename)
+        if m != None:
+            filename = m.group(1)
```

Comment:
This is the point where I'm not sure about what is the best solution. The issue is that after URL grabbing we obtain links like this: `http://site.com/filename.txt#some-long-anchor?query=short` and we want to get rid of the part after the `#`.

In fact, I did it with `clean_url` function: https://github.com/coursera-dl/coursera-dl/blob/master/coursera/utils.py#L153

However I use it during the parsing stage:
https://github.com/coursera-dl/coursera-dl/blob/master/coursera/api.py#L275
https://github.com/coursera-dl/coursera-dl/blob/master/coursera/api.py#L670

The question is: when do we apply that function? After parsing or before downloading?

Also, as far I know, your test case should work with current code:
```
input: ml-classification/08_boosting/04_programming-assignment-1/01_exploring-ensemble-methods_login.jsp?tp=&arnumber=5128907&url=http%3A%2F%2Fieeexplore.ieee.org%2Fiel5%2F69%2F5173046%2F05128907.pdf%3farnumber%3d5128907
output: ml-classification/08_boosting/04_programming-assignment-1/01_exploring-ensemble-methods_login.jsp
```
I think we strip URLs better now.